packages/dds/tree/src/simple-tree/toMapTree.ts

```diff
@@ -236,94 +182,122 @@ function mapValueWithFallbacks(
-			const childSchema = schema.getFieldSchema(brand(key));
-			const mappedField = fieldDataToMapTrees(value, globalSchema, childSchema);
-			fields.set(brand(key), mappedField);
+			const mappedField = nodeDataToMapTree(value, childSchema);
+			fields.set(brand(key), [mappedField]);
 		}
 	}
 	return {
-		type: schema.name,
+		type: brand(schema.identifier),
 		fields,
 	};
 }
 
-function recordToMapTree(
+function objectToMapTree(
 	data: Record<string | number | symbol, InsertableContent>,
-	globalSchema: FlexTreeSchema,
-	typeSet: AllowedTypeSet,
+	allowedTypes: AllowedTypes,
 ): MapTree {
-	const schema = getType(data, globalSchema, typeSet);
+	const schema = getType(data, allowedTypes);
+	if (schema.kind !== NodeKind.Object) {
+		fail(`Provided object input is incompatible with schema "${schema.identifier}".`);
+	}
 
 	const fields = new Map<FieldKey, MapTree[]>();
 
 	// Filter keys to only those that are strings - our trees do not support symbol or numeric property keys
 	const keys = Reflect.ownKeys(data).filter((key) => typeof key === "string") as FieldKey[];
 
 	for (const key of keys) {
-		assert(!fields.has(key), 0x84d /* Keys should not be duplicated */);
-		const value = data[key];
+		const fieldValue = data[key];
 
 		// Omit undefined record entries - an entry with an undefined key is equivalent to no entry
-		if (value !== undefined) {
-			const childSchema = schema.getFieldSchema(key);
-			const mappedChildTree = fieldDataToMapTrees(value, globalSchema, childSchema);
-			fields.set(brand(key), mappedChildTree);
+		if (fieldValue !== undefined) {
+			const fieldSchema = getObjectFieldSchema(schema, key);
+			const mappedChildTree = nodeDataToMapTree(fieldValue, fieldSchema.allowedTypes);
+			fields.set(brand(key), [mappedChildTree]);
 		}
 	}
 
 	return {
-		type: schema.name,
+		type: brand(schema.identifier),
 		fields,
 	};
 }
 
-function getType(
-	data: InsertableContent,
-	globalSchema: FlexTreeSchema,
-	typeSet: AllowedTypeSet,
-): FlexTreeNodeSchema {
-	const possibleTypes = getPossibleTypes(
-		globalSchema,
-		typeSet,
-		data as ContextuallyTypedNodeData,
-	);
+function getObjectFieldSchema(schema: TreeNodeSchema, key: FieldKey): NormalizedFieldSchema {
+	assert(schema.kind === NodeKind.Object, "Expected an object schema.");
+	const fields = schema.info as Record<string, ImplicitFieldSchema>;
+	if (fields[key] === undefined) {
+		fail(`Field "${key}" not found in schema "${schema.identifier}".`);
+	} else {
+		return normalizeFieldSchema(fields[key]);
+	}
+}
+
+function getType(data: InsertableContent, allowedTypes: AllowedTypes): TreeNodeSchema {
+	const possibleTypes = getPossibleTypes(allowedTypes, data as ContextuallyTypedNodeData);
+	if (possibleTypes.length === 0) {
+		throw new UsageError(
+			`The provided data is incompatible with all of the types allowed by the schema. The set of allowed types is: ${JSON.stringify(
+				[...allowedTypes.map((schema) => evaluateLazySchema(schema).identifier)],
+				undefined,
```

Comment:
I think this explicit undefined parameter is needless as that's the default

Reply:
I actually have no idea where that came from. Possible copy-paste. Possibly copilot. Will remove.